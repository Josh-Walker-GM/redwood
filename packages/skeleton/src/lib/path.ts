import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
