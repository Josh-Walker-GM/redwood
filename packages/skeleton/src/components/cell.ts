import fs from 'fs'
import path from 'path'

import traverse from '@babel/traverse'
import type { ExportNamedDeclaration } from '@babel/types'
import {
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/internal/dist/paths'

import { getASTFromCode } from '../lib/ast'
import { getGraphQLQueryName } from '../lib/gql'

import { RedwoodSkeleton } from './base'
import type { RedwoodProject } from './project'

export class RedwoodCell extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly hasQueryExport: boolean

  readonly gqlQuery: string | undefined
  readonly gqlQueryName: string | undefined

  readonly hasLoadingExport: boolean
  readonly hasEmptyExport: boolean
  readonly hasFailureExport: boolean
  readonly hasSuccessExport: boolean

  readonly isValid: boolean

  constructor(filepath: string) {
    super(filepath)

    const code = fs.readFileSync(this.filepath, { encoding: 'utf8', flag: 'r' })
    const ast = getASTFromCode(code)

    const namedExports: ExportNamedDeclaration[] = []
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        namedExports.push(path.node)
      },
    })

    const namedExportsNames = namedExports
      .map((node) => {
        if (
          isVariableDeclaration(node.declaration) &&
          isVariableDeclarator(node.declaration.declarations[0]) &&
          isIdentifier(node.declaration.declarations[0].id)
        ) {
          return node.declaration.declarations[0].id.name
        }
        return false
      })
      .filter(Boolean)

    // Check for a cells expected export fields
    this.hasQueryExport = namedExportsNames.includes('QUERY')
    this.hasLoadingExport = namedExportsNames.includes('Loading')
    this.hasEmptyExport = namedExportsNames.includes('Empty')
    this.hasFailureExport = namedExportsNames.includes('Failure')
    this.hasSuccessExport = namedExportsNames.includes('Success')

    // Check if the mandatory success export is present
    if (!this.hasSuccessExport) {
      this.errors.push('No "Success" export found but one is required')
    }

    // Check if the mandatory query export is present
    if (!this.hasQueryExport) {
      this.errors.push('No "QUERY" export found but one is required')
    } else {
      const graphqlExport = namedExports.find((node) => {
        if (
          node.declaration != null &&
          isVariableDeclaration(node.declaration) &&
          isVariableDeclarator(node.declaration.declarations[0]) &&
          isIdentifier(node.declaration.declarations[0].id)
        ) {
          return node.declaration.declarations[0].id.name === 'QUERY'
        }
        return false
      }) as ExportNamedDeclaration // Note: This should not be undefined given this.hasQueryExport == true

      if (
        graphqlExport.declaration != null &&
        isVariableDeclaration(graphqlExport.declaration) &&
        isVariableDeclarator(graphqlExport.declaration.declarations[0]) &&
        graphqlExport.declaration.declarations[0].init != null
      ) {
        const gqlVariable = graphqlExport.declaration.declarations[0].init
        switch (gqlVariable.type) {
          case 'TaggedTemplateExpression':
            if (gqlVariable.quasi.start && gqlVariable.quasi.end) {
              const graphQLSource = code.substring(
                gqlVariable.quasi.start + 1,
                gqlVariable.quasi.end - 1
              )
              this.gqlQuery = graphQLSource
              this.gqlQueryName = getGraphQLQueryName(this.gqlQuery)
              if (this.gqlQueryName === undefined) {
                this.errors.push('Could not determine the GraphQL query name')
              }
            } else {
              this.errors.push(
                `Could not extract the GraphQL query from "${gqlVariable.type}"`
              )
            }
            break
          default:
            this.errors.push(
              `Could not process the GraphQL query from "${gqlVariable.type}"`
            )
            break
        }
      } else {
        this.errors.push('Could not process the GraphQL variable')
      }
    }

    // Determine if the cell is valid
    this.isValid = this.hasQueryExport && this.hasSuccessExport
  }

  // Diagnostics

  getStatistics(): string {
    throw new Error('Method not implemented.')
  }
}

export function extractCell(filepath: string): RedwoodCell {
  return new RedwoodCell(filepath)
}

export function extractCells(
  project: RedwoodProject | undefined = undefined
): RedwoodCell[] {
  const cells: RedwoodCell[] = []

  const componentsPath = project
    ? getPaths(project.filepath).web.components
    : getPaths().web.components

  if (!fs.existsSync(componentsPath)) {
    return cells
  }

  // TODO: Confirm this the condition to detect a cell
  // Cells must be defined within a file which ends with `Cell.{js, jsx, tsx}`
  const getCellFiles = (directory: string) => {
    const cellFiles: string[] = []

    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isDirectory()) {
        cellFiles.push(...getCellFiles(path.join(directory, content)))
      } else if (stat.isFile()) {
        if (content.match(/.+Cell\.(js|jsx|tsx)$/)) {
          cellFiles.push(path.join(directory, content))
        }
      }
    })

    return cellFiles
  }

  const cellFiles = getCellFiles(componentsPath)
  cellFiles.forEach((cellPath) => {
    cells.push(extractCell(cellPath))
  })

  return cells
}
