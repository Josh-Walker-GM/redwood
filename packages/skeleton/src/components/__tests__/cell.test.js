import path from 'path'

import { getCells } from '../cell'

describe('From inside the test-project fixture', () => {
  const FIXTURE_PATH = path.join(
    __dirname,
    '..',
    '..',
    '..',
    '..',
    '..',
    '__fixtures__',
    'test-project'
  )

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = FIXTURE_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  describe('without project', () => {
    it('returns the correct cells', () => {
      const cells = getCells()
      expect(cells).toBe({})
    })
  })
})
