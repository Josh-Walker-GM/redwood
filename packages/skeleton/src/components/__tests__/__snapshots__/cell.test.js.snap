// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture returns the correct cells with a project: cell count 1`] = `0`;

exports[`From within the empty-project fixture returns the correct cells without a project: cell count 1`] = `0`;

exports[`From within the example-todo-main fixture returns the correct cells with a project: /web/src/components/NumTodosCell/NumTodosCell.js 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/NumTodosCell/NumTodosCell.js",
  "gqlQuery": "
  query NumTodosCell_GetCount {
    todosCount
  }
",
  "gqlQueryName": "NumTodosCell_GetCount",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "NumTodosCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells with a project: /web/src/components/NumTodosTwoCell/NumTodosTwoCell.js 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/NumTodosTwoCell/NumTodosTwoCell.js",
  "gqlQuery": "
  query NumTodosCell_GetCount {
    todosCount
  }
",
  "gqlQueryName": "NumTodosCell_GetCount",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "NumTodosTwoCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells with a project: /web/src/components/TableCell/TableCell.js 1`] = `
RedwoodCell {
  "errors": [
    "No "Success" export found but one is required",
    "No "QUERY" export found but one is required",
  ],
  "filepath": "/web/src/components/TableCell/TableCell.js",
  "gqlQuery": undefined,
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": false,
  "hasQueryExport": false,
  "hasSuccessExport": false,
  "isValid": false,
  "name": "TableCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells with a project: /web/src/components/TodoListCell/TodoListCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/TodoListCell/TodoListCell.tsx",
  "gqlQuery": "
  query TodoListCell_GetTodos {
    todos {
      id
      body
      status
    }
  }
",
  "gqlQueryName": "TodoListCell_GetTodos",
  "hasAfterQueryExport": true,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "TodoListCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells with a project: cell count 1`] = `4`;

exports[`From within the example-todo-main fixture returns the correct cells without a project: /web/src/components/NumTodosCell/NumTodosCell.js 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/NumTodosCell/NumTodosCell.js",
  "gqlQuery": "
  query NumTodosCell_GetCount {
    todosCount
  }
",
  "gqlQueryName": "NumTodosCell_GetCount",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "NumTodosCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells without a project: /web/src/components/NumTodosTwoCell/NumTodosTwoCell.js 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/NumTodosTwoCell/NumTodosTwoCell.js",
  "gqlQuery": "
  query NumTodosCell_GetCount {
    todosCount
  }
",
  "gqlQueryName": "NumTodosCell_GetCount",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "NumTodosTwoCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells without a project: /web/src/components/TableCell/TableCell.js 1`] = `
RedwoodCell {
  "errors": [
    "No "Success" export found but one is required",
    "No "QUERY" export found but one is required",
  ],
  "filepath": "/web/src/components/TableCell/TableCell.js",
  "gqlQuery": undefined,
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": false,
  "hasQueryExport": false,
  "hasSuccessExport": false,
  "isValid": false,
  "name": "TableCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells without a project: /web/src/components/TodoListCell/TodoListCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/TodoListCell/TodoListCell.tsx",
  "gqlQuery": "
  query TodoListCell_GetTodos {
    todos {
      id
      body
      status
    }
  }
",
  "gqlQueryName": "TodoListCell_GetTodos",
  "hasAfterQueryExport": true,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "TodoListCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct cells without a project: cell count 1`] = `4`;

exports[`From within the example-todo-main-with-errors fixture returns the correct cells with a project: /web/src/components/TableCell/TableCell.js 1`] = `
RedwoodCell {
  "errors": [
    "No "Success" export found but one is required",
    "No "QUERY" export found but one is required",
  ],
  "filepath": "/web/src/components/TableCell/TableCell.js",
  "gqlQuery": undefined,
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": false,
  "hasQueryExport": false,
  "hasSuccessExport": false,
  "isValid": false,
  "name": "TableCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct cells with a project: /web/src/components/TodoListCell/TodoListCell.js 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/TodoListCell/TodoListCell.js",
  "gqlQuery": "
  {
    todos {
      id
      body
      status
    }
  }
",
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "TodoListCell",
  "warnings": [
    "Could not determine the GraphQL query name",
  ],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct cells with a project: cell count 1`] = `2`;

exports[`From within the example-todo-main-with-errors fixture returns the correct cells without a project: /web/src/components/TableCell/TableCell.js 1`] = `
RedwoodCell {
  "errors": [
    "No "Success" export found but one is required",
    "No "QUERY" export found but one is required",
  ],
  "filepath": "/web/src/components/TableCell/TableCell.js",
  "gqlQuery": undefined,
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": false,
  "hasQueryExport": false,
  "hasSuccessExport": false,
  "isValid": false,
  "name": "TableCell",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct cells without a project: /web/src/components/TodoListCell/TodoListCell.js 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/TodoListCell/TodoListCell.js",
  "gqlQuery": "
  {
    todos {
      id
      body
      status
    }
  }
",
  "gqlQueryName": undefined,
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": false,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "TodoListCell",
  "warnings": [
    "Could not determine the GraphQL query name",
  ],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct cells without a project: cell count 1`] = `2`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/AuthorCell/AuthorCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/AuthorCell/AuthorCell.tsx",
  "gqlQuery": "
  query FindAuthorQuery($id: Int!) {
    author: user(id: $id) {
      email
      fullName
    }
  }
",
  "gqlQueryName": "FindAuthorQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "AuthorCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/BlogPostCell/BlogPostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/BlogPostCell/BlogPostCell.tsx",
  "gqlQuery": "
  query FindBlogPostQuery($id: Int!) {
    blogPost: post(id: $id) {
      id
      title
      body
      author {
        email
        fullName
      }
      createdAt
    }
  }
",
  "gqlQueryName": "FindBlogPostQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "BlogPostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/BlogPostsCell/BlogPostsCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/BlogPostsCell/BlogPostsCell.tsx",
  "gqlQuery": "
  query BlogPostsQuery {
    blogPosts: posts {
      id
      title
      body
      author {
        email
        fullName
      }
      createdAt
    }
  }
",
  "gqlQueryName": "BlogPostsQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "BlogPostsCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/Contact/ContactCell/ContactCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Contact/ContactCell/ContactCell.tsx",
  "gqlQuery": "
  query FindContactById($id: Int!) {
    contact: contact(id: $id) {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "FindContactById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "ContactCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/Contact/ContactsCell/ContactsCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Contact/ContactsCell/ContactsCell.tsx",
  "gqlQuery": "
  query FindContacts {
    contacts {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "FindContacts",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "ContactsCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/Contact/EditContactCell/EditContactCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Contact/EditContactCell/EditContactCell.tsx",
  "gqlQuery": "
  query EditContactById($id: Int!) {
    contact: contact(id: $id) {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "EditContactById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "EditContactCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/Post/EditPostCell/EditPostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Post/EditPostCell/EditPostCell.tsx",
  "gqlQuery": "
  query EditPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "EditPostById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "EditPostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/Post/PostCell/PostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Post/PostCell/PostCell.tsx",
  "gqlQuery": "
  query FindPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindPostById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "PostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/Post/PostsCell/PostsCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Post/PostsCell/PostsCell.tsx",
  "gqlQuery": "
  query FindPosts {
    posts {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindPosts",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "PostsCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: /web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx",
  "gqlQuery": "
  query FindWaterfallBlogPostQuery($id: Int!) {
    waterfallBlogPost: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindWaterfallBlogPostQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "WaterfallBlogPostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells with a project: cell count 1`] = `10`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/AuthorCell/AuthorCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/AuthorCell/AuthorCell.tsx",
  "gqlQuery": "
  query FindAuthorQuery($id: Int!) {
    author: user(id: $id) {
      email
      fullName
    }
  }
",
  "gqlQueryName": "FindAuthorQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "AuthorCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/BlogPostCell/BlogPostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/BlogPostCell/BlogPostCell.tsx",
  "gqlQuery": "
  query FindBlogPostQuery($id: Int!) {
    blogPost: post(id: $id) {
      id
      title
      body
      author {
        email
        fullName
      }
      createdAt
    }
  }
",
  "gqlQueryName": "FindBlogPostQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "BlogPostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/BlogPostsCell/BlogPostsCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/BlogPostsCell/BlogPostsCell.tsx",
  "gqlQuery": "
  query BlogPostsQuery {
    blogPosts: posts {
      id
      title
      body
      author {
        email
        fullName
      }
      createdAt
    }
  }
",
  "gqlQueryName": "BlogPostsQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "BlogPostsCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/Contact/ContactCell/ContactCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Contact/ContactCell/ContactCell.tsx",
  "gqlQuery": "
  query FindContactById($id: Int!) {
    contact: contact(id: $id) {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "FindContactById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "ContactCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/Contact/ContactsCell/ContactsCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Contact/ContactsCell/ContactsCell.tsx",
  "gqlQuery": "
  query FindContacts {
    contacts {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "FindContacts",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "ContactsCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/Contact/EditContactCell/EditContactCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Contact/EditContactCell/EditContactCell.tsx",
  "gqlQuery": "
  query EditContactById($id: Int!) {
    contact: contact(id: $id) {
      id
      name
      email
      message
      createdAt
    }
  }
",
  "gqlQueryName": "EditContactById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "EditContactCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/Post/EditPostCell/EditPostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Post/EditPostCell/EditPostCell.tsx",
  "gqlQuery": "
  query EditPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "EditPostById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": false,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "EditPostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/Post/PostCell/PostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Post/PostCell/PostCell.tsx",
  "gqlQuery": "
  query FindPostById($id: Int!) {
    post: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindPostById",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "PostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/Post/PostsCell/PostsCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/Post/PostsCell/PostsCell.tsx",
  "gqlQuery": "
  query FindPosts {
    posts {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindPosts",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "PostsCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: /web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx 1`] = `
RedwoodCell {
  "errors": [],
  "filepath": "/web/src/components/WaterfallBlogPostCell/WaterfallBlogPostCell.tsx",
  "gqlQuery": "
  query FindWaterfallBlogPostQuery($id: Int!) {
    waterfallBlogPost: post(id: $id) {
      id
      title
      body
      authorId
      createdAt
    }
  }
",
  "gqlQueryName": "FindWaterfallBlogPostQuery",
  "hasAfterQueryExport": false,
  "hasBeforeQueryExport": false,
  "hasEmptyExport": true,
  "hasFailureExport": true,
  "hasIsEmptyExport": false,
  "hasLoadingExport": true,
  "hasQueryExport": true,
  "hasSuccessExport": true,
  "isValid": true,
  "name": "WaterfallBlogPostCell",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct cells without a project: cell count 1`] = `10`;
